Framework/Algorithms/src/ReflectometryReductionOne2.cpp

```diff
@@ -1,18 +1,261 @@
+    // an operation (+-:) that produces even more groups.
+    auto groups = Kernel::StringTokenizer(
+        instructions, ",",
+        StringTokenizer::TOK_TRIM | StringTokenizer::TOK_IGNORE_EMPTY);
+    for (const auto &groupStr : groups) {
+      // Look for the various operators in the string. If one is found then
+      // do the necessary translation into groupings.
+      if (groupStr.find('+') != std::string::npos) {
+        // add a group with the given spectra
+        translateAdd(groupStr, outGroups);
+      } else if (groupStr.find('-') != std::string::npos) {
+        translateSumRange(groupStr, outGroups);
+      } else if (groupStr.find(':') != std::string::npos) {
+        translateRange(groupStr, outGroups);
+      } else if (!groupStr.empty()) {
+        // contains no instructions, just add this spectrum as a new group
+        // create group of size 1 with the spectrum in it and add it to output
+        outGroups.emplace_back(1, boost::lexical_cast<size_t>(groupStr));
+      }
+    }
+  } catch (boost::bad_lexical_cast &) {
+    throw std::runtime_error("Invalid processing instructions: " +
+                             instructions);
+  }
+
+  return outGroups;
+}
+
+/**
+* Map a spectrum index from the given map to the given workspace
+* @param map : the original spectrum number to index map
+* @param mapIdx : the index to look up in the map
+* @param destWS : the destination workspace
+*/
+int mapSpectrumIndexToWorkspace(const spec2index_map &map, const size_t mapIdx,
+                                MatrixWorkspace_const_sptr destWS) {
+
+  // Get the spectrum numbers for these indices
+  auto it = std::find_if(map.begin(), map.end(),
+                         [mapIdx](const std::pair<specnum_t, size_t> &mapIter) {
+                           return mapIter.second == mapIdx;
+                         });
+
+  if (it == map.end()) {
+    std::ostringstream errMsg;
+    errMsg << "Workspace index " << mapIdx << " not found in run workspace ";
+    throw std::runtime_error(errMsg.str());
+  }
+  specnum_t specId = it->first;
+
+  // Get the workspace index in the destination workspace
+  int wsIdx = static_cast<int>(destWS->getIndexFromSpectrumNumber(specId));
+
+  return wsIdx;
+}
+
+/**
+* @param originWS : Origin workspace, which provides the original workspace
+* index to spectrum number mapping.
+* @param hostWS : Workspace onto which the resulting workspace indexes will be
+* hosted
+* @throws :: If the specId are not found to exist on the host end-point
+*workspace.
+* @return :: Remapped workspace indexes applicable for the host workspace,
+*as a vector of groups of vectors of spectrum indices
+*/
+std::vector<std::vector<size_t>> mapSpectrumIndicesToWorkspace(
+    MatrixWorkspace_const_sptr originWS, MatrixWorkspace_const_sptr hostWS,
+    const std::vector<std::vector<size_t>> &detectorGroups) {
+
+  auto map = originWS->getSpectrumToWorkspaceIndexMap();
```

Comment:
Can you check with Simon or Owen, is using the spectrum number to workspace index map is still valid? I am not sure if it will be dropped at some point.